src/systemd-sonic-generator/systemd-sonic-generator.c

```diff
@@ -0,0 +1,299 @@
+        unit_files[num_unit_files] = strdup(line);
+        num_unit_files++;
+    }
+
+    free(line);
+
+    fclose(fp);
+
+    return num_unit_files;
+}
+
+
+static int install_unit_file(char* unit_file, char* target, char* install_dir) {
+    /***
+    Creates a symlink for a unit file installation
+
+    For a given unit file and target directory,
+    create the appropriate symlink in the target directory
+    to enable the unit and have it started by Systemd
+    ***/
+    char final_install_dir[PATH_MAX];
+    char src_path[PATH_MAX];
+    char dest_path[PATH_MAX];
+    struct stat st;
+    int r;
+
+    assert(unit_file);
+    assert(target);
+
+    strcpy(final_install_dir, install_dir);
+    strcat(final_install_dir, target);
+
+    strcpy(src_path, UNIT_FILE_PREFIX);
+    strcat(src_path, unit_file);
+
+    if (stat(final_install_dir, &st) == -1) {
+        // If doesn't exist, create
+        r = mkdir(final_install_dir, 0755);
+        if (r == -1) {
+            fputs("Unable to create target directory\n", stderr);
+            return -1;
+        }
+    }
+    else if (S_ISREG(st.st_mode)) {
+        // If is regular file, remove and create
+        r = remove(final_install_dir);
+        if (r == -1) {
+            fputs("Unable to remove file with same name as target directory\n", stderr);
+            return -1;
+        }
+
+        r = mkdir(final_install_dir, 0755);
+        if (r == -1) {
+            fputs("Unable to create target direectory\n", stderr);
+            return -1;
+        }
+    }  
+    else if (S_ISDIR(st.st_mode)) {
+        // If directory, verify correct permissions
+        r = chmod(final_install_dir, 0755);
+        if (r == -1) {
+            fputs("Unable to change permissions of existing target directory\n", stderr);
+            return -1;
+        }
+    }
+        
+
+    strcpy(dest_path, final_install_dir);
+    strcat(dest_path, "/");
+    strcat(dest_path, unit_file);
+
+    r = symlink(src_path, dest_path);
+
+    if (r < 0) {
+        if (errno == EEXIST)
+            return 0;
+
+        return 1;
+    }
+
+    return 0;
+}
+
+
+int main(int argc, char **argv) {
+    char* unit_files[MAX_NUM_UNITS];
+    char install_dir[PATH_MAX];
+    char* targets[MAX_NUM_TARGETS];
+    int num_unit_files;
+    int num_targets;
+
+    if (argc <= 1) {
+        fputs("Installation directory required as argument\n", stderr);
+        return 1;
+    }
+
+    strcpy(install_dir, argv[1]);
+    strcat(install_dir, "/");
+
+    num_unit_files = get_unit_files(unit_files);
+
+    // For each unit file, get the installation targets and install the unit
+    for (int i = 0; i < num_unit_files; i++) {
+        num_targets = get_install_targets(unit_files[i], targets);
+        if (num_targets < 0)
+            return 1;
+
+        for (int j = 0; j < num_targets; j++) {
+            if (install_unit_file(unit_files[i], targets[j], install_dir) != 0) {
+                return 1;
```

Comment:
it is better to print error and continue to next target, you do not want to fail the whole process just because one bad unit file.